### ISSUE TITLE
SATT 2.0/PB - Changes to Bundle contents not reflected in subsequent renewals

### ISSUE THREAD

**Issue by sambomcd** [Prospress/woocommerce-subscribe-all-the-things#160]
**Use case:**
- Subscribed to a Product Bundle type product
- Change the contents of the Bundle (i.e. add/remove products within the bundle)
- Subsequent renewals still reflect the original bundle contents from when the subscriptions started. It does not update to reflect the new composition of the Product Bundle

**Expected behaviour**: subsequent renewals reflect any changes made to the Product Bundle contents between renewals.

(**Disclaimer:** I am not 100% certain that the 'expected behaviour' above is actually the way it is supposed to work, but it is how I assumed it to work, and I think how it worked pre-2.0!)

**Comment by sambomcd** [comment]
(I am just commenting to cancel my earlier 'disclaimer'!) Having given this more thought, I think even if it wasn't the way it worked pre-2.0, it seems to me that this refreshing of Product Bundle contents is logically how it needs to function. Subscribing to a Product Bundle is not the same as subscribing to each of the individual bundled products within the Bundle, which is the current behaviour.

**Comment by franticpsyx** [comment]
Thanks for the follow-up @sambomcd - closing!

**Comment by sambomcd** [comment]
@franticpsyx I'm not sure why you are closing? I was confirming in my follow up that the behaviour is not working as it should?

**Comment by franticpsyx** [comment]
@sambomcd There are still a couple loose ends with Product Bundles & Composite Products compatibility that affect renewals/re-subscriptions -- these need to be patched on the PB/CP side.

Please drop me a line over at support@somewherewarm.gr and I'll see what I can do.

**Comment by franticpsyx** [comment]
> it seems to me that this refreshing of Product Bundle contents is logically how it needs to function

For posterity, the "right" behavior IMO is this:

If the contents of an existing bundle change, then effectively:

- the bundle configuration might be changed, 
- some bundled products may no longer exist, or 
- some new ones may have been added.

If so, then the behavior will be the same as when attempting to renew/resubscribe a subscription that contains invalid products.

@thenbrent I am not sure if there are any cases where renewals are handled exclusively on the gateway side?

@sambomcd For "grandfathering" to work correctly, you'd need to keep the existing bundles intact but perhaps hide them in your catalog, and create new ones to include the changes you have in mind.

**Comment by sambomcd** [comment]
Hmmm. I can see it working both ways. In my current situation, the Product Bundle is a food bag, with recipes that get swapped each week. The customer subscribes to the Bag itself, expecting the contents to be different each week. The subscription is tied to the Bag as that is the constant, and where the price is set. This is not a 'grandfathering' situation, per se. We are not, logically, bringing in new bundles to replace old. It is the same bundle.

Appreciate that may not be the normal use case - but it is ours!

**Comment by thenbrent** [comment]
> @thenbrent I am not sure if there are any cases where renewals are handled exclusively on the gateway side?

Yep there are. PayPal Standard, WorldPay and any other gateway which doesn't support _Recurring Total Modifications_, as listed here: https://docs.woocommerce.com/document/subscriptions/payment-gateways/#section-3

> For posterity, the "right" behavior IMO is this

@franticpsyx the behaviour you describe is the default behaviour by Subscriptions for all products. Modifying a product's details doesn't change subscriptions with that product as a line item.

You may want to add a flag of some form to allow bundle updates to be reflected in subscriptions, there's certainly many valid use cases I've seen over the years for that type of behaviour over the years, but there are a lot of gotchas in doing that. It's not easy thing.

It is often also equivalent to a "bulk edit" feature, which is in some ways how we'll be dealing with similar issues in Subscriptions itself eventually.

**Comment by franticpsyx** [comment]
> the behaviour you describe is the default behaviour by Subscriptions for all products. Modifying a product's details doesn't change subscriptions with that product as a line item.

@thenbrent -- "right" was probably a bad term to use. I was thinking exactly what you said, which is better described as "technically more feasible" from a dev's perspective and "more common" from an end-use perspective.

> Hmmm. I can see it working both ways. In my current situation, the Product Bundle is a food bag, with recipes that get swapped each week.

@sambomcd Thanks for sharing that! This is definitely a valid use case that I would throw in our "subscription content switching" bucket. See the [Roadmap](https://github.com/Prospress/woocommerce-subscribe-all-the-things/wiki/Roadmap) for details. 

As @thenbrent mentioned, we will probably need to deal with quite a few "challenges" when we start looking into scheme/content switching in SATT. It's too early to imagine how this will/should work in different applications (e.g. content switching initiated by the customer vs content switching initiated by the store owner). For now all I can say is that there's a long path we need to walk before we start looking into switching-related features.

**Comment by sambomcd** [comment]
Appreciate the input and explanations, guys. 

So what I am understanding is:

- the way I thought it worked by default is wrong
- the way it works by default is the complete opposite of what I thought
- making it behave in the way I describe is possible, but complicated

You guys sure know how to destroy a man's hopes and dreams! ;)

This is a bit tricky, as my use case seems (to me at least) not far fetched.  I can easily imagine other scenarios, for example where a vendor might wish to offer a subscription to a Weekly or Monthly Specials package. The customer would sign up expecting that the contents are going to change each time.

Thanks for sharing the Roadmap (how had I not seen that). Good to know this will go into the mix for consideration.

In case it helps to have the specifics of a real-life use case for when you get to this, I am specifically trying to leverage Subscriptions+SATT+PB to allow:

- bundling of 'Recipe' products (containing a PDF for the recipe as a download) into a Food bag product
- ability to purchase a Food Bag either as a one of or as an ongoing subscription (impossible without SATT)
- have emails go out to customers each week automatically with the links to that week's Recipes

The last one is where the 'refresh' of the contents is critical. At the moment, they just keep getting the same recipes as were part of the Food Bag when they first purchased the Sub.

 I have to come up with another plan, then. I need to provide this sort of functionality now. :(

**Comment by thenbrent** [comment]
> You guys sure know how to destroy a man's hopes and dreams! ;)

Sorry! :( It will eventually be possible, just not yet.

The good news, or at least, not-so-bad news is that from a code perspective, it's not crazy hard to use different line items on each renewal order.

The hard part is creating the actual interface for that.

While the **Edit Product** interface could be used in the case of a Bundled product, this problem is actually more generic than just Product Bundles. For a generic plugin, it should be possible for other product types too. For example, as you say, it could be used for a weekly or monthly specials box. The best way to do this would be to use a _Variable_ product with _Renewal Period_ as an attribute, so _weekly_ and _monthly_ become options for the customer to choose from. The **Edit Product** screen for _Variable Products_ has no UI that could be used to specify which items to provide on the next renewal for that. So we need to build that as part of it. It would very likely look similar to the _Bundled Products_ tab and allow you to choose products, but it would not be exclusively for the _Product Bundle_ product type.

This is a long way of saying that if you wanted this built custom, you could actually use the existing _Product Bundle_ interface and just have the code to use those items on each renewal order. If you want recommendations on freelancers or agencies who can help with that let me know.

> In case it helps to have the specifics of a real-life use case

It always helps! Thanks for sharing.

**Comment by sambomcd** [comment]
Thanks for the comments and reassurances @thenbrent!! I appreciate you guys have a lot on, and you are doing a fantastic job with things like SATT that I am not complaining about this. (PS Are there still plans for SATT become how the Subscriptions plugin works by default?? Subscriptions on its own would not be much use to me!)

I also appreciate your idea around how I could approach doing this myself. I actually had an idea for a simpler (dirtier) workaround temporarily. As I really only need the Bundle contents refreshed to send out an email to customers, I might look instead at customising the Processing/Renewal Processing emails. I am thinking I could get the Product Bundle info from the Subscription, then go and look up the current bundled products and list those out in the mail.  I could also hopefully grab the Download link (to the PDF file for each recipe) and include that in the mail, although I think this will run into problems with the permissions for Downloads.

Anyway, not a pretty solution, but it might do until you and the team are able to get to this.

Many thanks again for taking the time to share your thoughts on this.

**Comment by thenbrent** [comment]
> Are there still plans for SATT become how the Subscriptions plugin works by default??

Yes definitely. That's the final item in the _Long Term_ section of the roadmap: https://github.com/Prospress/woocommerce-subscribe-all-the-things/wiki/Roadmap

And the more in-depth background/explanation is here: https://github.com/Prospress/woocommerce-subscribe-all-the-things/wiki/Future-of-Subscribe-All-the-Things
